util/workgraph/workgraph.go

```diff
@@ -0,0 +1,449 @@
+}
+
+// AddNode adds a new Node to the WorkGraph with the provided options. It
+// returns an error if the given Node.ID was already added to the WorkGraph, or
+// if one of the options provided was invalid.
+func (g *WorkGraph) AddNode(n Node, opts *AddNodeOpts) error {
+	id := n.ID()
+	if _, found := g.nodes[id]; found {
+		return fmt.Errorf("node %q already exists", id)
+	}
+	g.nodes[id] = n
+
+	if opts == nil {
+		return nil
+	}
+
+	// Create an edge from each dependency pointing to this node, forcing
+	// that node to be evaluated first.
+	for _, dep := range opts.Dependencies {
+		if _, found := g.nodes[dep]; !found {
+			return fmt.Errorf("dependency %q not found", dep)
+		}
+		g.edges.Add(dep, id)
+	}
+	return nil
+}
+
+// MustAddNode calls AddNode and panics on error.
+func (g *WorkGraph) MustAddNode(n Node, opts *AddNodeOpts) {
+	if err := g.AddNode(n, opts); err != nil {
+		panic(err)
+	}
+}
+
+// Run will iterate through all Nodes in this WorkGraph, running them once all
+// their dependencies have been satisfied, and returning any errors that occur.
+func (g *WorkGraph) Run(ctx context.Context) error {
+	groups, err := g.topoSortKahn()
+	if err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	type queueEntry struct {
+		id   string
+		done chan struct{}
+	}
+
+	// Create one goroutine that pushes jobs onto our queue...
+	var wg sync.WaitGroup
+	queue := make(chan queueEntry)
+	publishCtx, publishCancel := context.WithCancel(ctx)
+	defer publishCancel()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		defer close(queue)
+
+		// For each parallel group...
+		var dones []chan struct{}
+		for _, group := range groups {
+			dones = dones[:0] // re-use existing storage, if any
+
+			// Push all items in this group onto our queue
+			for curr := range group {
+				done := make(chan struct{})
+				dones = append(dones, done)
+
+				select {
+				case <-publishCtx.Done():
+					return
+				case queue <- queueEntry{curr, done}:
+				}
+			}
+
+			// Now that we've started everything, wait for them all
+			// to complete.
+			for _, done := range dones {
+				select {
+				case <-publishCtx.Done():
+					return
+				case <-done:
+				}
+			}
+
+			// Now that we've done this entire group, we can
+			// continue with the next one.
+		}
+	}()
+
+	firstErr := make(chan error, 1)
+	saveErr := func(err error) {
+		if err == nil {
+			return
+		}
+
+		// Tell the publisher to shut down
+		publishCancel()
+
+		select {
+		case firstErr <- err:
+		default:
+		}
+	}
+
+	// ... and N goroutines that each work on an item from the queue.
+	n := g.Concurrency
+	if n == 0 {
+		n = runtime.GOMAXPROCS(-1)
+	}
+
+	wg.Add(n)
+	for i := 0; i < n; i++ {
+		go func() {
+			defer wg.Done()
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case ent, ok := <-queue:
+					if !ok {
+						return
+					}
+
+					// immediately call a function so we
+					// can handle panics
+					err := func() (retErr error) {
```

Comment:
could probably extract this out into its own func, i don't think it closes on anything?